### ISSUE TITLE
FR: Progress Sync should use metadata as primary document identifier when available, and file checksum as backup due to file conversion impacting file checksum

### ISSUE THREAD

**Issue by IcedEagle** [koreader/koreader#7836]
Currently, Progress Sync uses an MD5 checksum over 10x1k bytes of the file, with uneven reads weighted towards the header of the file (.../frontend/document/document.lua: function Document:fastDigest). This works well considering how many file formats KOReader supports. However, if your file management software, such as Calibre performs any sort of file conversion to make the file better suited to your device, the file checksum is likely to break and progress sync will not recognize the files on each device as being the same book. While you could tell the user never to do any file conversion, this ignores the fact that the file conversion is desirable to some users as it reduces file size among other optimizations. I think Progress Sync should be made more resilient to potential file conversion artifacts impacting its ability to sync what the user perceives as the same document across devices.

**Describe the solution you'd like**
When Progress Sync requests the documents identifier, there should be a check to see if the document's format provides any readable metadata to use as a unique identifier. If so, use that identifier as defined in the file format specification. If there is no suitable unique identifier or metadata available, then fall back to the current file-based MD5 hash algorithm.

For a specific example, on the ePub 2 specification, per the OPF spec portion, each package is required to have a unique identifier defined. I propose that when we detect this is an ePub 2 document, we use the document's OPF defined unique identifier, and run that through the same MD5 hash algorithm to give us our new ID. Using this method, even if the user has multiple devices, and a uniquely converted book on each of those devices, as long as the conversion software doesn't modify the unique identifier, progress sync will work as the user expects.

Whatever identifier is used on the client continue to be run through MD5, with the resulting hash string used as the output to ensure there is no sync-server change necessary.

**Describe alternatives you've considered**
I have not gone through all supported file formats to understand how many file formats have uniquely defined identifiers, so it's not clear how many formats would benefit from this. However, ePUB format would benefit. Because I propose we use the current method as a fall back, it means other document format support for this could be added over time, and they would not be impacted initially.

**Additional context**
Some consideration needs to be made on how to handle existing progress sync for users. KOReader performs the current checksum when first opening a file and caches it to disk, so any existing devices would continue to sync if we changed the document identifier used. Any device of a user's that would open a file for the first time would not have matching identifier's as an older device though.

**Comment by Frenzie** [comment]
> For a specific example, on the ePub 2 specification, per the OPF spec portion, each package is required to have a unique identifier defined. I propose that when we detect this is an ePub 2 document, we use the document's OPF defined unique identifier, and run that through the same MD5 hash algorithm to give us our new ID. Using this method, even if the user has multiple devices, and a uniquely converted book on each of those devices, as long as the conversion software doesn't modify the unique identifier, progress sync will work as the user expects.

Minor sidenote, but couldn't that just as easily be changed by Calibre/Sigil/etc.?

**Comment by IcedEagle** [comment]
Yes. I do call that out saying that as long as the conversion software doesn't modify that value it would work.

However, if that value were to be changed by any of the conversion/editing software today, it also would have a high chance to break the current file reading checksum as well. Either way you end up with broken progress sync, so I don't believe it would make it any less resilient than the current method used. With the current versions of Calibre that I have tried, doing conversions of the same book does not modify the unique identifier, preserving OPF spec intent of that field.

**Comment by Frenzie** [comment]
Format-specific PRs are welcome by me in any case. I was just wondering if it might not be wasted effort for whoever might be interested. ;-)

**Comment by poire-z** [comment]
> However, if your file management software, such as Calibre performs any sort of file conversion to make the file better suited to your device, the file checksum is likely to break and progress sync will not recognize the files on each device as being the same book. 

> For a specific example, on the ePub 2 specification, per the OPF spec portion, each package is required to have a unique identifier defined

What kind of tweak to an EPUB can calibre do? I don't really see what it could do to make it better adapt to some device. But if it tweaks the content/styling, I'm afraid it should no longer be considered the same book for KOReader: the DOM and xpaths/xpointers (for bookmarks and last read position) may then change, and if you use/sync metadata.epub.lua from different documents, things won't map and it'll be a mess. So, even for EPUB, a checksum feels safer: you're then sure it's the exact same document.

**Comment by IcedEagle** [comment]
For device specific customizations, I'm not sure the full extent of what Calibre offers. The primary one I use has Calibre scale down any images to better fit my device to shrink file space of each book. I don't need massive images which are higher res than the screen itself after all! I think it also does a gray-scale conversion, whereas on my Android it leaves it full color, and full resolution.

Also, in my case Calibre will re-order some fields in the OPF metadata file when sending to my Kobo for no reason, and white space is slightly different. Probably a bug for Calibre, but since the content is the same and as far as I can tell, it still meets the OPF spec, I wouldn't mark it as a major issue either, other than it breaks KOReader's Progress Sync :).

I don't know of or use any device specific customizations that would touch the DOM of the book itself. Everything I use only touches the images and metadata. However, that doesn't mean there aren't plug-ins out there that do that. I do agree that using file checksum is safer in being certain it is the exact same file with a high degree of certainty. Presumably there is some error handling around bookmarks and last read positions if they're invalid, but it is possible this change could expose edge cases in that handling in extreme cases.

As an alternative, instead of using the UUID in the ePub metadata maybe we could still perform a checksum but only over the Text XML content documents for an ePub? That way, it would prevent any changes around images, metadata, fonts, from impacting the resulting checksum, while ensuring that we would be able to maintain bookmarks, last read positions and such? Then, if a user does have device specific plug-ins that change content drastically, it wouldn't be any different to today.

**Comment by poire-z** [comment]
I'll let calibre users tell more about how they manage that.
But it feels excessive having to hack our rendering engines to help with such use cases (_has Calibre scale down any images to better fit my device to shrink file space_ ) :)

And if it's just about KOSync, well, let's handle that on its side: why not add an option so it computes the checksum from the filename instead of the content ? :)

**Comment by NiLuJe** [comment]
@poire-z raises a good point, as a *full* calibre conversion will essentially spit out something *completely* different.

What happens on "Send To Device", though, is something *different* (with a caveat for ePub -> KePub implicit conversions, but that should, by all means, be disabled by the user when using KOReader).

TL;DR: While the proposal makes sense, I foresee a high number of possibly weird corner-cases.

That said: I kinda still view possibly-kinda-off-sync in some corner-cases *better* than no-sync-because-hash-mismatch?

**Comment by IcedEagle** [comment]
I was able to get a dev environment working, and I decided to implement this using the suggestion from @poire-z which made it considerably easier to start. Basically, I provide a user option to switch to using filename for the checksum, instead of binary, which remains the default.

I did some amount of testing locally to see how KOReader would react if I named two completely separate books with the same filename, and how it would react in various situations. It seemed to react okay from my perspective. It would ask to sync progress, and if the user confirmed they wanted to sync, it would bring them back to the beginning of the book as the sync point doesn't exist in this other book.

My draft PR is here: https://github.com/koreader/koreader/pull/7840

(Apologies if I'm doing any of this process incorrectly. First time using git and working with koreader so learning a lot here...)

My concern is whether filename is too loose and might run into too many false positives. I suppose we'd be weighing the false positives to this, versus the amount of false negatives today with the binary method, which goes into the point @NiLuJe makes. I'd agree in general I would prefer a false positive to a false negative, but only to a certain extent. At the very least, I feel like if a user runs into this situation, it's easier to have them rename a file to stop the false positive, versus having them figure out how to get the files to be binary identical to stop the false negative.

Additionally, maybe it would make more sense to have the user set a default, but be able to override it on a per-document basis if necessary. It might be overkill though. In my case, I would set it to filename-based and leave it.

**Comment by poire-z** [comment]
(I'm not a calibre or KOSync user.)
Looks ok and simple enough to me. If it solves your own use case, that's good enough. No need to overthink it until somebody complains and wants more :)

**Comment by craig-sanders** [comment]
> And if it's just about KOSync, well, let's handle that on its side: why not add an option so it computes the checksum from the filename instead of the content ? :)

Using the filename is useful in some contexts I guess, but doesn't always work.  A filename is not unique nor is it **consistent across devices**. Solving this kind of inconsistency is part of the reason why Universally Unique IDentifiers, or UUIDs, exist.

Here's an example of why filenames don't always work:

I have koreader installed on my Linux desktop, configured to read books directly from my calibre libraries under `~/ebooks/` (which is not ideal, but I can't see a way around it without having yet another copy of the books just for koreader, which is even less ideal).  All of my ebooks are epub, either natively or converted with Calibre.  Anyway, books in that directory are saved as `~/ebooks/Library/Author/Series/book name (calibre index)/title - author.epub`. The base filename is `title - author.epub`.

I transfer epubs to my android phone and tablet devices using calibre.  Calibre is configured to use `{author_sort[0]:uppercase()}/{author_sort}/{series}/{series_index:0>3s||.}{title}` as the filename template.  The base filename is just `title.epub`, which does not match `title - author.epub`.

I do not wish to change the template for several reasons, one of which is that I do not want to have to delete and re-transfer several thousand books and lose over a decade's worth of my reading history in the process.

Without any way to do automatic syncing, my current "process" is to remember which device I last read a book on, check the chapter number/title there, and manually change to that on the new device (this is especially a PITA on books which only have a "Start" entry in the TOC and no chapters, my only recourse is to edit the epub and fix the TOC). And manually keep each book in sync by opening them on each device.  Far from ideal.

Using the UUID from the epub's metadata if it exists would solve that problem.  The uuid can be found in the `dc:identifier opf:scheme="uuid" id="uuid_id"` tag, e.g.:

```
        <dc:identifier opf:scheme="uuid" id="uuid_id">xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx</dc:identifier>
```

Alternatively, calibre also adds its own uuid:

```
<dc:identifier opf:scheme="calibre">xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx</dc:identifier>
```

Either of these would work. Probably best to use the `scheme="uuid"` one if it exists, otherwise use the `scheme="calibre"` uuid if that exists.  There's no need to hash either of these with md5 as they're already unique.

The metadata can be found in `OEBPS/content.opf` in the epub itself, and in the separate `metadata.opf` file in the same directory as the epub in a calibre library.

Of course, if metadata UUIDs don't exist for a given book, then fall back to using the filename or md5 hash or whatever.

I have no opinion on whether this should be the default setting or not, but it should be available as an option.

---

PS: as for the earlier suggestion to hash only the .xhtml files, that would work for most books where the contents never change, but wouldn't work for web serial books downloaded from sites like Royal Road with tools like the Calibre FanFicFare plugin (or its standalone version). Every time there's a new chapter, the md5 hash would change.

Also, why waste CPU power (and battery) hashing .epub file contents when there's a perfectly good UUID just waiting to be used?

Technically, you could say "but it changed, it's not the same book".  Practically, this is how anyone reading a web serial as an epub would expect it to work.

**Comment by Svampebob1** [comment]
What about a combination? Use UUID as primary identifier. If UUID is unavailable for the book or the UUID does not match the stored identifier for the book, fall back to MD5.

**Comment by craig-sanders** [comment]
> What about a combination? Use UUID as primary identifier. If UUID is unavailable for the book or the UUID does not match the stored identifier for the book, fall back to MD5.

Makes sense to me.   And I still think that using an epub's UUID is the optimal and the technically correct method, that's why there's a UUID in the spec.

I ended up using filenames, wrote a perl script to create a symlink farm from my Calibre libraries so that koreader could open the exact same filenames on my desktop PC as on my tablet and phone.   And another perl script to use with the "Open With" calibre plugin so I can open one or more epubs in that symlink farm with koreader just by pressing `k`.   Not perfect, but a **lot** better than nothing.

Both perl scripts are hard-coded for my book filename template - I did not want to write an interpreter for calibre's template language. Nor did I wish to write it in python because calibre's source code is hundreds of files worth of uncommented, undocumented spaghetti code. It probably makes perfect sense to the author, but anyone else would have to spend a year or two studying it to understand it.

**Comment by Frenzie** [comment]
Like I said. ;-)

> Format-specific PRs are welcome by me in any case.

Edit: albeit potentially less welcome than five years ago due to stochastic parrots, but that's still fine as long as it's properly edited down by the human.

**Comment by Svampebob1** [comment]
I'm looking into this now.. I'm not sure other formats than epub contains UUID. 

If you read a book using epub on one device and the same book using AZW3 or PDF on a different device, both UUID and MD5 will probably not match.

Please correct me if I'm wrong 😊.

Edit: If anything, we will have another method  of identifying a book, if the calculation of MD5 from epub has a mismatch between devices.

**Comment by Frenzie** [comment]
> If you read a book using epub on one device and the same book using AZW3 or PDF on a different device, both UUID and MD5 will probably not match.

AZW3 isn't supported.

But that aside, if the UUID *did* match, it'd be problematic because by definition reading progress can't match between different document types (except maybe between HTML and XHTML).

**Comment by Svampebob1** [comment]
> > If you read a book using epub on one device and the same book using AZW3 or PDF on a different device, both UUID and MD5 will probably not match.
> 
> AZW3 isn't supported.
> 
> But that aside, if the UUID _did_ match, it'd be problematic because by definition reading progress can't match between different document types (except maybe between HTML and XHTML).

I haven't used the plugin for very long, so I haven't gotten the time testing it thoroughly enough. Does the current implementation support sync of reading progress between different document types?

**Comment by Frenzie** [comment]
PDF is a type of paged media. Progress means you're on page 15.

EPUB progress means something like `/html/body/DocFragment[6250]/body/h1[33]/text().0`.

The two aren't compatible, and even if they were (e.g., EPUB and FB2 use the same type of progress mechanism) they'd be irreconcilable.

**Comment by Svampebob1** [comment]
> PDF is a type of paged media. Progress means you're on page 15.
> 
> EPUB progress means something like `/html/body/DocFragment[6250]/body/h1[33]/text().0`.
> 
> The two aren't compatible, and even if they were (e.g., EPUB and FB2 use the same type of progress mechanism) they'd be irreconcilable.

Thanks for the insight.
